pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -249,6 +249,10 @@ def __init__(
         self.__validate_init_configuration()
         self._save_function = None
 
+    @property
+    def state_id(self) -> str:
+        return self._generate_state_id(monitor=self.monitor)
```

Comment:
Users will definitely want more attributes added to this.

Should we review all the possible attributes and decide? Did you only choose monitor by default for a reason?

Same for `EarlyStopping`

Reply:
I chose monitor because that's the most discriminative option for these callbacks. 
The rest are to be chosen for more granularity. 

YES: the ones where we are positive they need to be included for strictness when reloading
NO: the ones that are completely irrelevant for state OR where we are certain users want to change them when reloading.
MAYBE: I am not sure yet.

Here is my selection:

**EarlyStopping**

```py
        monitor: Optional[str] = None,  # YES
        min_delta: float = 0.0,    # NO
        patience: int = 3,    # NO
        verbose: bool = False,  # NO
        mode: str = "min",  # YES
        strict: bool = True,  # NO
        check_finite: bool = True,   # NO
        stopping_threshold: Optional[float] = None,    # NO
        divergence_threshold: Optional[float] = None,    # NO
        check_on_train_epoch_end: bool = True,    # MAYBE
```

**ModelCheckpoint**
```py

        dirpath: Optional[Union[str, Path]] = None,    # NO
        filename: Optional[str] = None,   # NO
        monitor: Optional[str] = None,   # YES
        verbose: bool = False,   # NO
        save_last: Optional[bool] = None,   # NO
        save_top_k: int = 1,      # YES
        save_weights_only: bool = False,   # NO
        mode: str = "min",    # YES
        auto_insert_metric_name: bool = True,   # NO
        every_n_train_steps: Optional[int] = None,  # MAYBE
        train_time_interval: Optional[timedelta] = None,   # MAYBE
        every_n_epochs: Optional[int] = None,  # MAYBE
        save_on_train_epoch_end: Optional[bool] = None,   # MAYBE
        period: Optional[int] = None,   # NO
        every_n_val_epochs: Optional[int] = None,  # NO
```

Reply:
>         check_on_train_epoch_end: bool = True,    # MAYBE

So I'd say "no" for this because I can't think of a use case for having 2 `EarlyStopping` callbacks with the same monitor but different `check_on_train_epoch_end` values.

>        save_last: Optional[bool] = None,   # NO
>        every_n_train_steps: Optional[int] = None,  # MAYBE
>        train_time_interval: Optional[timedelta] = None,   # MAYBE
>        every_n_epochs: Optional[int] = None,  # MAYBE

I'd say "yes" for all of these.

>         save_on_train_epoch_end: Optional[bool] = None,   # MAYBE

In this case, I'm leaning towards yes because having a `monitor` is not required for `MoelCheckpoint`.
I can imagine somebody wanting to with `monitor=None` both `on_train_epoch_end` true and false.

>        period: Optional[int] = None,   # NO
>        every_n_val_epochs: Optional[int] = None,  # NO

These are deprecated so surely "no".

We can chat offline for further discussions about each

Reply:
@carmocca for the choices of interval settings like every_n_train_steps, train_time_interval, every_n_epochs you said "yes" because you think a user wants to have several version of a callback with different intervals. But that will also put the restriction that these settings can't be changed at the time of resuming anymore. Are we good with this?

Reply:
That is unfortunate but to me, better to have that constraint than to lose the state for those callbacks.

Reply:
Okay, to summarize. For the ModelCheckpoint we want to support this:

```python
callbacks=[
    ModelCheckpoint(every_n_train_steps=10),
    ModelCheckpoint(train_time_interval=1hour),
    ModelCheckpoint(every_n_epochs=5),
]
```
so we will add these attributes to the key in order to differentiate the states. We need to do so because the model checkpoint supports `monitor=None`.